src/abstractdataframe/join.jl

```diff
@@ -91,6 +91,291 @@ _rename_cols(old_names::AbstractVector{Symbol},
+end
+
+function compose_inner_table(joiner::DataFrameJoiner,
+                             makeunique::Bool,
+                             left_rename::Union{Function, AbstractString, Symbol},
+                             right_rename::Union{Function, AbstractString, Symbol})
+
+    right_shorter = length(joiner.dfr_on[!, 1]) <= length(joiner.dfl_on[!, 1])
+
+    left_cols = collect(eachcol(joiner.dfl_on))
+    right_cols = collect(eachcol(joiner.dfr_on))
+
+    disallow_sorted = false
+
+    if right_shorter
+        for i in eachindex(left_cols, right_cols)
+            rc = right_cols[i]
+            lc = left_cols[i]
+            if lc isa PooledArray
+                right_cols[i] = map2refs(rc, lc)
+                left_cols[i] = lc.refs
+            end
+        end
+    else
+        for i in eachindex(left_cols, right_cols)
+            rc = right_cols[i]
+            lc = left_cols[i]
+            if rc isa PooledArray
+                right_cols[i] = rc.refs
+                left_cols[i] = map2refs(lc, rc)
+            end
+
+            # this is a workaround for https://github.com/JuliaData/CategoricalArrays.jl/issues/319
+            rct = typeof(right_cols[i])
+            lct = typeof(left_cols[i])
+            rcat = rct.name === :CategoricalArray && nameof(rct.module) === :CategoricalArrays
+            lcat = lct.name === :CategoricalArray && nameof(lct.module) === :CategoricalArrays
+            disallow_sorted |= rcat ⊻ lcat
+        end
+    end
+
+    # TODO: if PooledArrays are found potentially the following optimizations can be done:
+    # 1. identify rows in shorter table that should be dropped
+    # 2. develop custom _innerjoin_sorted and _innerjoin_unsorted that
+    #    drop rows from shorter table that do not match rows from longer table based on
+    #    PooledArray refpool check
+    # this optimization significantly complicates the code (especially sorted path).
+    # It should be added if in practice we find that the use case is often enough
+    # and that the benefits are significant. The two cases when the benefits should
+    # be expected are:
+    # 1. Shorter table is sorted when we drop rows not matching longer table rows
+    # 2. Shorter table does not have duplicates when we drop rows not matching longer table rows
+
+    left_col = prepare_on_col(left_cols...)
+    right_col = prepare_on_col(right_cols...)
+
+    if isempty(left_col) || isempty(right_col)
+        # we treat this case separately so we know we have at least one element later
+        left_ixs, right_ixs = Int[], Int[]
+    else
+        both_sorted = false
+        if disallow_sorted
+            try
+                # the isless, isequal and isconcretetype tests are to make sure
+                # that if we use the fast path for sorted vectors we do not hit
+                # the problem that some entries are not comparable
+                isequal(left_col[1], right_col[1])
+                isless(left_col[1], right_col[1])
```

Comment:
these tests have to be done on the first non-missing value and `isconretetype` below probably should subtract `Missing` from the type.

Reply:
Sounds good. Though I was thinking that another option could be to add `try... catch` to `_innerjoin_sorted`, and fall back to the unsorted method if that fails. If some values cannot be compared, the failure is likely to happen quite soon, right?

Reply:
Yes - we can do this. We still have to exclude `CategoricalVector` case manually.